Chip-8 Challenge/src/CHIP-8 Virtual Machine/Keypad.cs

```diff
@@ -8,71 +8,81 @@ namespace CHIP_8_Virtual_Machine
 {
     public class Keypad
     {
-        private Dictionary<string, Nibble> _map = new Dictionary<string, Nibble>();
-        private Dictionary<Nibble, bool> _state = new Dictionary<Nibble, bool>();
+        private bool[] _state = new bool[16];
+        private Dictionary<string, Nibble> _map = new();
 
-        public bool IsDown(Nibble key)
-        {
-            if (_state.TryGetValue(key, out bool value))
-            {
-                return value;
-            }
+        public bool this[Nibble key] => _state[key];
 
-            return false;
+        public event EventHandler<Nibble> OnKeyDown;
+        public event EventHandler<Nibble> OnKeyUp;
+
+        public void WindowsKeyDown(char key)
+        {
+            WindowsKeyDown(key.ToString());
         }
 
-        public void KeyDown(string windowsKey)
+        public void WindowsKeyDown(string windowsKeyName)
         {
-            if (_map.ContainsKey(windowsKey))
+            if (_map.ContainsKey(windowsKeyName))
             {
-                Nibble key = _map[windowsKey];
+                Nibble key = _map[windowsKeyName];
                 _state[key] = true;
+                OnKeyDown?.Invoke(this, key);
             }
             else
             {
-                Console.WriteLine($"Key {windowsKey} not found in mapping");
+                throw new KeyNotFoundException($"Key {windowsKeyName} not found in mapping");
```

Comment:
We can debate this one today. Last week the crew decided that we should just ignore any keys outside the range as the physical Chip 8 keyboard wouldn't have any keys outside the range. Ignore but log. I would prefer handling at the calling level but was overruled.

Reply:
Preaching to the converted :-)